src/pytest_plugins/forks/forks.py

```diff
@@ -552,151 +560,354 @@ def fork(request):
+        return resulting_set
+
+
+class ValidUntil(ValidityMarker):
+    """
+    Marker to specify the fork until which the test is valid. The test will not be filled for
+    forks after the specified fork.
+
+    ```python
+    import pytest
+
+    from ethereum_test_tools import Alloc, StateTestFiller
+
+    @pytest.mark.valid_until("London")
+    def test_something_only_valid_until_london(
+        state_test: StateTestFiller,
+        pre: Alloc
+    ):
+        pass
+    ```
+
+    In this example, the test will only be filled for the London fork and before, e.g. London,
+    Berlin, Istanbul, etc.
+    """
+
+    def _process_with_marker_args(self, *fork_args) -> Set[Fork]:
+        """Process the fork arguments."""
+        forks: Set[Fork] = self.process_fork_arguments(*fork_args)
+        resulting_set: Set[Fork] = set()
+        for fork in forks:
+            resulting_set |= {f for f in self.all_forks if f <= fork}
+        return resulting_set
+
+
+class ValidAtTransitionTo(ValidityMarker, mutually_exclusive=True):
+    """
+    Marker to specify that a test is only meant to be filled at the transition to the specified
+    fork.
+
+    The test usually starts at the fork prior to the specified fork at genesis and at block 5 (for
+    pre-merge forks) or at timestamp 15,000 (for post-merge forks) the fork transition occurs.
+
+    ```python
+    import pytest
+
+    from ethereum_test_tools import Alloc, BlockchainTestFiller
+
+    @pytest.mark.valid_at_transition_to("London")
+    def test_something_that_happens_during_the_fork_transition_to_london(
+        blockchain_test: BlockchainTestFiller,
+        pre: Alloc
+    ):
+        pass
+    ```
+
+    In this example, the test will only be filled for the fork that transitions to London at block
+    number 5, `BerlinToLondonAt5`, and no other forks.
+
+    To see or add a new transition fork, see the `ethereum_test_forks.forks.transition` module.
+
+    Note that the test uses a `BlockchainTestFiller` fixture instead of a `StateTestFiller`,
+    as the transition forks are used to test changes throughout the blockchain progression, and
+    not just the state change of a single transaction.
+
+    This marker also accepts the following keyword arguments:
+
+    - `subsequent_transitions`: Force the test to also fill for subsequent fork transitions.
+    - `until`: Implies `subsequent_transitions` and puts a limit on which transition fork will the
+        test filling will be limited to.
+
+    For example:
+    ```python
+    @pytest.mark.valid_at_transition_to("Cancun", subsequent_transitions=True)
+    ```
+
+    produces tests on `ShanghaiToCancunAtTime15k` and `CancunToPragueAtTime15k`, and any transition
+    fork after that.
+
+    And:
+    ```python
+    @pytest.mark.valid_at_transition_to("Cancun", subsequent_transitions=True, until="Prague")
+    ```
+
+    produces tests on `ShanghaiToCancunAtTime15k` and `CancunToPragueAtTime15k`, but no forks after
```

Comment:
what would it mean the test is produced at transition?  what are the input arguments to such test function? 
transition test comes as a blockchain test with a few block before transition, a transition block and a few blocks after. in each block we put some test logic.  so it is   x-1, x, x + 1 at least. how would it look like in python?

Reply:
Same as a normal blockchain test, with the difference that you need to add enough blocks to reach the `15_000` timestamp, which can happen even at block 1.

```python
@pytest.mark.valid_at_transition_to("Prague")
    def test_something_that_happens_during_the_fork_transition_to_prague(
        blockchain_test: BlockchainTestFiller,
        pre: Alloc
    ):
        blocks = [
           Block(timestamp=1_000, tx=[...]),  # Cancun block
           Block(timestamp=2_000, tx=[...]),  # Cancun block
           ...
           Block(timestamp=15_000, tx=[...]),  # First Prague block
           Block(timestamp=16_000, tx=[...]),  # Second Prague block
```